File: sdk/dart/dovahlink_client/lib/src/protocol/protocol_revision_validator.dart
import 'package:dovahlink_client_sdk/src/protocol/protocol_format_exception.dart';

/// Validates revision fields shared by state snapshots and state events.
class ProtocolRevisionValidator {
  /// Rejects [value] when it is not an integer greater than or equal to zero.
  ///
  /// [fieldName] is included in the safe protocol-format diagnostic.
  ///
  /// @param value The decoded JSON value to validate.
  /// @param fieldName The protocol field name used in the diagnostic.
  static void validateNonNegativeInt(Object? value, String fieldName) {
    if (value is! int || value < 0) {
      throw ProtocolFormatException(
        '$fieldName must be a non-negative integer.',
      );
    }
  }
}

File: integration/DovahLinkValidationClient/ProtocolRevisionValidator.cs
namespace DovahLinkValidationClient;

/// <summary>
/// Validates decoded revision fields, shared by StateSnapshotPayload and StateEventPayload.
/// <c>protocol/schema/README.md</c>'s state envelope describes every revision field as a
/// non-negative integer; <see cref="System.Text.Json.Nodes.JsonNode.GetValue{T}"/> already rejects
/// a non-integer JSON value, so this only needs to enforce the sign.
/// </summary>
public static class ProtocolRevisionValidator
{
    /// <summary>
    /// Validates that a decoded revision value is non-negative.
    /// </summary>
    /// <param name="value">The already-decoded revision value.</param>
    /// <param name="fieldName">The field's name, for the exception message.</param>
    /// <exception cref="FormatException">Thrown when the value is negative.</exception>
    public static void ValidateNonNegative(long value, string fieldName)
    {
        if (value < 0)
        {
            throw new FormatException($"{fieldName} must be a non-negative integer.");
        }
    }
}
